src/transaction/log_recovery_redo.hpp

```diff
@@ -0,0 +1,594 @@
+    const LOG_REC_MVCC_UNDOREDO &log_rec, log_rcv &rcv, log_rectype log_rtype, struct log_zip &undo_unzip_support,
+    struct log_zip &redo_unzip_support)
+{
+  return log_rv_get_log_rec_redo_data<LOG_REC_UNDOREDO> (thread_p, log_pgptr_reader, log_rec.undoredo,
+	 rcv, log_rtype, undo_unzip_support, redo_unzip_support);
+}
+
+template <>
+inline int log_rv_get_log_rec_redo_data<LOG_REC_MVCC_REDO> (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+    const LOG_REC_MVCC_REDO &log_rec, log_rcv &rcv, log_rectype log_rtype, struct log_zip &undo_unzip_support,
+    struct log_zip &redo_unzip_support)
+{
+  return log_rv_get_unzip_and_diff_redo_log_data (thread_p, log_pgptr_reader, &rcv, 0, nullptr, redo_unzip_support);
+}
+
+template <>
+inline int log_rv_get_log_rec_redo_data<LOG_REC_REDO> (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+    const LOG_REC_REDO &log_rec, log_rcv &rcv, log_rectype log_rtype, struct log_zip &undo_unzip_support,
+    struct log_zip &redo_unzip_support)
+{
+  return log_rv_get_unzip_and_diff_redo_log_data (thread_p, log_pgptr_reader, &rcv, 0, nullptr, redo_unzip_support);
+}
+
+template <>
+inline int log_rv_get_log_rec_redo_data<LOG_REC_RUN_POSTPONE> (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+    const LOG_REC_RUN_POSTPONE &log_rec, log_rcv &rcv, log_rectype log_rtype, struct log_zip &undo_unzip_support,
+    struct log_zip &redo_unzip_support)
+{
+  return log_rv_get_unzip_and_diff_redo_log_data (thread_p, log_pgptr_reader, &rcv, 0, nullptr, redo_unzip_support);
+}
+
+template <>
+inline int log_rv_get_log_rec_redo_data<LOG_REC_COMPENSATE> (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader,
+    const LOG_REC_COMPENSATE &log_rec, log_rcv &rcv, log_rectype log_rtype, struct log_zip &undo_unzip_support,
+    struct log_zip &redo_unzip_support)
+{
+  return log_rv_get_unzip_and_diff_redo_log_data (thread_p, log_pgptr_reader, &rcv, 0, nullptr, redo_unzip_support);
+}
+
+template <typename T>
+void log_rv_redo_record_sync (THREAD_ENTRY *thread_p, log_reader &log_pgptr_reader, const T &log_rec,
+			      const VPID &rcv_vpid, const log_lsa &rcv_lsa, const LOG_LSA *end_redo_lsa, LOG_RECTYPE log_rtype,
+			      LOG_ZIP &undo_unzip_support, LOG_ZIP &redo_unzip_support)
+{
+  // bit of debug code to ensure that, should this code be executed asynchronously, within the same page,
+  // the lsa is everincreasing, thus, not altering the order in which it has been added to the log in the first place
+//#if !defined(NDEBUG)
+//  static std::mutex mtx;
+//  static std::map<std::pair<short, int32_t>, log_lsa> map_proper_vpid_lsa;
+//  std::lock_guard<std::mutex> lck (mtx);
+//  std::pair<short, int32_t> key {rcv_vpid.volid, rcv_vpid.pageid};
+//  const auto key_it =  map_proper_vpid_lsa.find (key);
+//  if (key_it != map_proper_vpid_lsa.end())
+//    {
+//      assert ((*key_it).second < rcv_lsa);
+//    }
+//  map_proper_vpid_lsa.emplace(key, rcv_lsa);
+//#endif
```

Comment:
This check seems that it will be useful to detect any inconsistencies that may come after additional changes in the future. It is better if it is active in debug mode to immediately detect inconsistencies.

The map should be cleared after recovery. It might be useful to move the code in a self-contained class. The class may be defined in a log_recovery_redo.cpp file.

Thoughts?

Reply:
Done